Load modelling dataset

In [88]:
# ============================================================
# Project paths
# ============================================================

from pathlib import Path
import pandas as pd
import numpy as np

PROJECT_DIR = (
    Path.home()
    / "Desktop"
    / "Arun"
    / "Arun"
    / "DBA-Walsh"
    / "SecondYr"
    / "Capstone"
    / "ML_Driven_TalentIdentification_In_Youth_Chess"
)

RAW_DIR = PROJECT_DIR / "data" / "raw"
PROCESSED_DIR = PROJECT_DIR / "data" / "processed"
OUTPUT_TABLES_DIR = PROJECT_DIR / "outputs" / "tables"
OUTPUT_FIGURES_DIR = PROJECT_DIR / "outputs" / "figures"

# Create output folders if they do not already exist
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT_TABLES_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT_FIGURES_DIR.mkdir(parents=True, exist_ok=True)

print("PROJECT_DIR:", PROJECT_DIR)
print("RAW_DIR:", RAW_DIR)
print("PROCESSED_DIR:", PROCESSED_DIR)
print("Raw folder exists:", RAW_DIR.exists())

PROJECT_DIR: /Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess
RAW_DIR: /Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess/data/raw
PROCESSED_DIR: /Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess/data/processed
Raw folder exists: True


load your two main files from data/raw

In [90]:
# ============================================================
# Load raw player sample and FIDE calculation-page data
# ============================================================

player_path = RAW_DIR / "player_sample_1000.csv"
calc_path = RAW_DIR / "fide_calculations_all_players_standard_clean_combined.csv"

player_df = pd.read_csv(player_path)
calc_df = pd.read_csv(calc_path)

player_df.columns = player_df.columns.str.strip()
calc_df.columns = calc_df.columns.str.strip()

print("Player sample shape:", player_df.shape)
print("Calculation data shape:", calc_df.shape)

print("\nPlayer columns:")
print(player_df.columns.tolist())

print("\nCalculation columns:")
print(calc_df.columns.tolist())

Player sample shape: (1000, 17)
Calculation data shape: (24204, 24)

Player columns:
['ID Number', 'Name', 'Fed', 'Sex', 'Title', 'Birth_Year', 'rating_start', 'rating_end', 'months_available', 'total_games_12m', 'avg_monthly_games', 'max_monthly_games', 'active_months_12m', 'inactive_months_12m', 'rating_gain_12m', 'age_approx_2026', 'rating_band']

Calculation columns:
['fide_id', 'period', 'rating_type', 'tournament_name', 'event_location_fed', 'event_start_date', 'event_end_date', 'event_rc', 'event_ro', 'event_score', 'event_games', 'event_chg', 'event_k', 'event_k_chg', 'opponent_name', 'opponent_title', 'opponent_rating', 'rating_difference_over_400_flag', 'opponent_fed', 'score', 'n', 'chg', 'k', 'k_chg']


In [91]:
# Final modelling dataset path
final_features_path = PROCESSED_DIR / "player_features_final.csv"

Basic checks

In [93]:
# ============================================================
# Updated paths for modelling notebook
# ============================================================

from pathlib import Path
import pandas as pd
import numpy as np

# New GitHub/project folder
PROJECT_DIR = Path.home() / "Desktop" / "Arun" / "Arun" / "DBA-Walsh" / "SecondYr" / "Capstone" / "ML_Driven_TalentIdentification_In_Youth_Chess"

# Old download folder where source CSVs currently exist
OLD_DATA_DIR = Path.home() / "Downloads" / "fide_history"

# Project subfolders
RAW_DIR = PROJECT_DIR / "data" / "raw"
PROCESSED_DIR = PROJECT_DIR / "data" / "processed"
INTERIM_DIR = PROJECT_DIR / "data" / "interim"
OUTPUT_TABLES_DIR = PROJECT_DIR / "outputs" / "tables"
OUTPUT_FIGURES_DIR = PROJECT_DIR / "outputs" / "figures"

# Create folders if missing
for folder in [RAW_DIR, PROCESSED_DIR, INTERIM_DIR, OUTPUT_TABLES_DIR, OUTPUT_FIGURES_DIR]:
    folder.mkdir(parents=True, exist_ok=True)

print("PROJECT_DIR exists:", PROJECT_DIR.exists())
print("OLD_DATA_DIR exists:", OLD_DATA_DIR.exists())
print("Processed folder:", PROCESSED_DIR)

PROJECT_DIR exists: True
OLD_DATA_DIR exists: True
Processed folder: /Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess/data/processed


In [94]:
# ============================================================
# Input files
# ============================================================

PLAYER_PATH = OLD_DATA_DIR / "player_sample_1000.csv"
CALC_PATH = OLD_DATA_DIR / "fide_calculations_all_players_standard_clean_combined.csv"

print("PLAYER_PATH exists:", PLAYER_PATH.exists(), PLAYER_PATH)
print("CALC_PATH exists:", CALC_PATH.exists(), CALC_PATH)

PLAYER_PATH exists: True /Users/arunkumar/Downloads/fide_history/player_sample_1000.csv
CALC_PATH exists: True /Users/arunkumar/Downloads/fide_history/fide_calculations_all_players_standard_clean_combined.csv


In [95]:
# ============================================================
# Load baseline player sample and opponent-level calculation data
# ============================================================

players = pd.read_csv(PLAYER_PATH)
calc = pd.read_csv(CALC_PATH)

players.columns = players.columns.str.strip()
calc.columns = calc.columns.str.strip()

print("Players shape:", players.shape)
print("Calculation records shape:", calc.shape)

print("\nPlayer columns:")
print(players.columns.tolist())

print("\nCalculation columns:")
print(calc.columns.tolist())

Players shape: (1000, 17)
Calculation records shape: (24204, 24)

Player columns:
['ID Number', 'Name', 'Fed', 'Sex', 'Title', 'Birth_Year', 'rating_start', 'rating_end', 'months_available', 'total_games_12m', 'avg_monthly_games', 'max_monthly_games', 'active_months_12m', 'inactive_months_12m', 'rating_gain_12m', 'age_approx_2026', 'rating_band']

Calculation columns:
['fide_id', 'period', 'rating_type', 'tournament_name', 'event_location_fed', 'event_start_date', 'event_end_date', 'event_rc', 'event_ro', 'event_score', 'event_games', 'event_chg', 'event_k', 'event_k_chg', 'opponent_name', 'opponent_title', 'opponent_rating', 'rating_difference_over_400_flag', 'opponent_fed', 'score', 'n', 'chg', 'k', 'k_chg']


In [96]:
# ============================================================
# Create model_df and save final modelling dataset
# ============================================================

import pandas as pd
import numpy as np
from pathlib import Path

# New project path
PROJECT_DIR = Path.home() / "Desktop" / "Arun" / "Arun" / "DBA-Walsh" / "SecondYr" / "Capstone" / "ML_Driven_TalentIdentification_In_Youth_Chess"

# Old data folder where current CSVs exist
OLD_DATA_DIR = Path.home() / "Downloads" / "fide_history"

# Project folders
PROCESSED_DIR = PROJECT_DIR / "data" / "processed"
OUTPUT_TABLES_DIR = PROJECT_DIR / "outputs" / "tables"

PROCESSED_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT_TABLES_DIR.mkdir(parents=True, exist_ok=True)

# Input files
PLAYER_PATH = OLD_DATA_DIR / "player_sample_1000.csv"
CALC_PATH = OLD_DATA_DIR / "fide_calculations_all_players_standard_clean_combined.csv"

# Load files
players = pd.read_csv(PLAYER_PATH)
calc = pd.read_csv(CALC_PATH)

players.columns = players.columns.str.strip()
calc.columns = calc.columns.str.strip()

# Standardize IDs
players["ID Number"] = players["ID Number"].astype(str).str.strip()
calc["fide_id"] = calc["fide_id"].astype(str).str.strip()

print("Players shape:", players.shape)
print("Calculation records shape:", calc.shape)
print("Unique players in sample:", players["ID Number"].nunique())
print("Unique players in calculation data:", calc["fide_id"].nunique())

Players shape: (1000, 17)
Calculation records shape: (24204, 24)
Unique players in sample: 1000
Unique players in calculation data: 629


In [97]:
# ============================================================
# Aggregate FIDE calculation-page records to player level
# ============================================================

calc_features = calc.copy()

# Convert possible numeric columns
numeric_candidates = [
    "opponent_rating",
    "rating",
    "score",
    "event_rc",
    "rc",
    "chg",
    "k_chg",
    "game_chg",
    "rating_change",
    "k_factor"
]

for col in numeric_candidates:
    if col in calc_features.columns:
        calc_features[col] = (
            calc_features[col]
            .astype(str)
            .str.replace("*", "", regex=False)
            .str.replace(",", "", regex=False)
        )
        calc_features[col] = pd.to_numeric(calc_features[col], errors="coerce")

# Detect usable columns
opponent_rating_col = "opponent_rating" if "opponent_rating" in calc_features.columns else None
if opponent_rating_col is None and "rating" in calc_features.columns:
    opponent_rating_col = "rating"

score_col = "score" if "score" in calc_features.columns else None

k_chg_col = None
for possible_col in ["k_chg", "chg", "rating_change", "game_chg"]:
    if possible_col in calc_features.columns:
        k_chg_col = possible_col
        break

print("Using opponent rating column:", opponent_rating_col)
print("Using score column:", score_col)
print("Using rating-change column:", k_chg_col)

# Build aggregation dictionary
agg_dict = {
    "calc_games_12m": ("fide_id", "count")
}

if "tournament_name" in calc_features.columns:
    agg_dict["calc_tournaments_12m"] = ("tournament_name", "nunique")

if opponent_rating_col is not None:
    agg_dict["avg_opponent_rating_12m"] = (opponent_rating_col, "mean")
    agg_dict["max_opponent_rating_12m"] = (opponent_rating_col, "max")
    agg_dict["games_vs_2000plus_12m"] = (opponent_rating_col, lambda x: (x >= 2000).sum())

if score_col is not None:
    agg_dict["total_score_12m"] = (score_col, "sum")

if k_chg_col is not None:
    agg_dict["total_k_chg_12m"] = (k_chg_col, "sum")
    agg_dict["avg_game_chg_12m"] = (k_chg_col, "mean")

calc_player_features = (
    calc_features
    .groupby("fide_id")
    .agg(**agg_dict)
    .reset_index()
)

# Score percentage
if "total_score_12m" in calc_player_features.columns:
    calc_player_features["score_percentage_12m"] = (
        calc_player_features["total_score_12m"] / calc_player_features["calc_games_12m"]
    )

print("Player-level calculation features shape:", calc_player_features.shape)
display(calc_player_features.head())

Using opponent rating column: opponent_rating
Using score column: score
Using rating-change column: k_chg
Player-level calculation features shape: (629, 10)


,fide_id,calc_games_12m,calc_tournaments_12m,avg_opponent_rating_12m,max_opponent_rating_12m,games_vs_2000plus_12m,total_score_12m,total_k_chg_12m,avg_game_chg_12m,score_percentage_12m
0,14350670,124,16,1581.822581,2013.0,2,39.0,-343.20,-2.767742,0.314516
1,25100122,8,1,1647.500000,1753.0,0,1.0,-32.80,-4.100000,0.125000
2,25100688,12,3,1610.666667,1869.0,0,1.0,0.00,0.000000,0.083333
3,25110462,18,2,1618.000000,1759.0,0,7.0,-16.00,-0.888889,0.388889
4,25113909,54,4,1592.481481,1984.0,0,29.0,10.26,0.190000,0.537037


In [98]:
# ============================================================
# Merge player sample with calculation-page features
# ============================================================

model_df = players.merge(
    calc_player_features,
    left_on="ID Number",
    right_on="fide_id",
    how="left"
)

# Flag calculation-page availability
model_df["has_calculation_data"] = model_df["calc_games_12m"].notna().astype(int)

# Fill count-based enrichment variables with 0
count_cols = [
    "calc_games_12m",
    "calc_tournaments_12m",
    "games_vs_2000plus_12m"
]

for col in count_cols:
    if col in model_df.columns:
        model_df[col] = model_df[col].fillna(0)

print("Final model_df shape:", model_df.shape)
print("\nCalculation data coverage:")
print(model_df["has_calculation_data"].value_counts(dropna=False))

display(model_df.head())

Final model_df shape: (1000, 28)

Calculation data coverage:
has_calculation_data
1    629
0    371
Name: count, dtype: int64


,ID Number,Name,Fed,Sex,Title,Birth_Year,rating_start,rating_end,months_available,total_games_12m,...,calc_games_12m,calc_tournaments_12m,avg_opponent_rating_12m,max_opponent_rating_12m,games_vs_2000plus_12m,total_score_12m,total_k_chg_12m,avg_game_chg_12m,score_percentage_12m,has_calculation_data
0,558016627,Jivtesh Tuli,IND,M,NaN,2011.0,1522,1502,3,8,...,0.0,0.0,NaN,NaN,0.0,NaN,NaN,NaN,NaN,0
1,531075452,Shukla Rajvi Mahesh,IND,F,NaN,2013.0,1606,1562,6,11,...,12.0,1.0,1460.000000,1515.0,0.0,2.0,0.0,0.000000,0.166667,1
2,537082728,Bishan Das,IND,M,NaN,2012.0,1449,1449,1,8,...,2.0,1.0,1777.000000,1777.0,0.0,0.0,0.0,0.000000,0.000000,1
3,25789783,Raman Siddhardh B,IND,M,NaN,2011.0,1847,1705,12,62,...,210.0,13.0,1698.790476,2286.0,14.0,129.0,163.2,0.777143,0.614286,1
4,45056404,Sradha Suman Jena,IND,F,NaN,2008.0,1482,1482,12,0,...,0.0,0.0,NaN,NaN,0.0,NaN,NaN,NaN,NaN,0


In [99]:
# ============================================================
# Save final modelling dataset
# ============================================================

FINAL_MODEL_PATH = PROCESSED_DIR / "player_features_final.csv"

model_df.to_csv(FINAL_MODEL_PATH, index=False)

print("Saved final modelling dataset to:")
print(FINAL_MODEL_PATH)
print("Final shape:", model_df.shape)

Saved final modelling dataset to:
/Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess/data/processed/player_features_final.csv
Final shape: (1000, 28)


In [100]:
print(model_df.columns.tolist())
print(model_df["has_calculation_data"].value_counts())

['ID Number', 'Name', 'Fed', 'Sex', 'Title', 'Birth_Year', 'rating_start', 'rating_end', 'months_available', 'total_games_12m', 'avg_monthly_games', 'max_monthly_games', 'active_months_12m', 'inactive_months_12m', 'rating_gain_12m', 'age_approx_2026', 'rating_band', 'fide_id', 'calc_games_12m', 'calc_tournaments_12m', 'avg_opponent_rating_12m', 'max_opponent_rating_12m', 'games_vs_2000plus_12m', 'total_score_12m', 'total_k_chg_12m', 'avg_game_chg_12m', 'score_percentage_12m', 'has_calculation_data']
has_calculation_data
1    629
0    371
Name: count, dtype: int64


Reload final modelling dataset

In [102]:
# ============================================================
# Step 1: Reload final modelling dataset
# ============================================================

from pathlib import Path
import pandas as pd
import numpy as np

PROJECT_DIR = Path.home() / "Desktop" / "Arun" / "Arun" / "DBA-Walsh" / "SecondYr" / "Capstone" / "ML_Driven_TalentIdentification_In_Youth_Chess"

PROCESSED_DIR = PROJECT_DIR / "data" / "processed"
OUTPUT_TABLES_DIR = PROJECT_DIR / "outputs" / "tables"
OUTPUT_TABLES_DIR.mkdir(parents=True, exist_ok=True)

FINAL_MODEL_PATH = PROCESSED_DIR / "player_features_final.csv"

model_df = pd.read_csv(FINAL_MODEL_PATH)
model_df.columns = model_df.columns.str.strip()

print("Loaded:", FINAL_MODEL_PATH)
print("Shape:", model_df.shape)

print("\nCalculation data coverage:")
print(model_df["has_calculation_data"].value_counts(dropna=False))

print("\nColumns:")
print(model_df.columns.tolist())

Loaded: /Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess/data/processed/player_features_final.csv
Shape: (1000, 28)

Calculation data coverage:
has_calculation_data
1    629
0    371
Name: count, dtype: int64

Columns:
['ID Number', 'Name', 'Fed', 'Sex', 'Title', 'Birth_Year', 'rating_start', 'rating_end', 'months_available', 'total_games_12m', 'avg_monthly_games', 'max_monthly_games', 'active_months_12m', 'inactive_months_12m', 'rating_gain_12m', 'age_approx_2026', 'rating_band', 'fide_id', 'calc_games_12m', 'calc_tournaments_12m', 'avg_opponent_rating_12m', 'max_opponent_rating_12m', 'games_vs_2000plus_12m', 'total_score_12m', 'total_k_chg_12m', 'avg_game_chg_12m', 'score_percentage_12m', 'has_calculation_data']


Confirm target and feature columns

In [104]:
# ============================================================
# Step 2: Confirm modelling columns
# ============================================================

target_col = "rating_gain_12m"

baseline_features = [
    "rating_start",
    "total_games_12m",
    "active_months_12m",
    "inactive_months_12m"
]

enriched_features = [
    "calc_games_12m",
    "calc_tournaments_12m",
    "avg_opponent_rating_12m",
    "score_percentage_12m",
    "games_vs_2000plus_12m",
    "total_k_chg_12m"
]

all_check_cols = [target_col] + baseline_features + enriched_features + ["has_calculation_data"]

check = pd.DataFrame({
    "column": all_check_cols,
    "exists": [col in model_df.columns for col in all_check_cols],
    "missing_count": [model_df[col].isna().sum() if col in model_df.columns else None for col in all_check_cols]
})

display(check)

,column,exists,missing_count
0,rating_gain_12m,True,0
1,rating_start,True,0
2,total_games_12m,True,0
3,active_months_12m,True,0
4,inactive_months_12m,True,0
5,calc_games_12m,True,0
6,calc_tournaments_12m,True,0
7,avg_opponent_rating_12m,True,371
8,score_percentage_12m,True,371
9,games_vs_2000plus_12m,True,0


RQ1 hypothesis testing

In [106]:
# ============================================================
# Step 3: RQ1 Hypothesis Test
# Current rating only vs multivariate baseline model
# ============================================================

import statsmodels.api as sm
from statsmodels.stats.anova import anova_lm

rq1_features = [
    "rating_start",
    "total_games_12m",
    "active_months_12m",
    "inactive_months_12m"
]

rq1_data = model_df[[target_col] + rq1_features].dropna().copy()

print("RQ1 sample size:", rq1_data.shape)

y = rq1_data[target_col]

# Null model: rating only
X_null = sm.add_constant(rq1_data[["rating_start"]])
model_null = sm.OLS(y, X_null).fit()

# Full model: rating + activity features
X_full = sm.add_constant(rq1_data[rq1_features])
model_full = sm.OLS(y, X_full).fit()

anova_rq1 = anova_lm(model_null, model_full)

print("RQ1 Null Model R-squared:", round(model_null.rsquared, 4))
print("RQ1 Full Model R-squared:", round(model_full.rsquared, 4))
print("RQ1 Full Model Adjusted R-squared:", round(model_full.rsquared_adj, 4))

print("\nRQ1 Nested Regression F-test:")
display(anova_rq1)

print("\nRQ1 Full Model Summary:")
print(model_full.summary())

RQ1 sample size: (1000, 5)
RQ1 Null Model R-squared: 0.0048
RQ1 Full Model R-squared: 0.2211
RQ1 Full Model Adjusted R-squared: 0.218

RQ1 Nested Regression F-test:


,df_resid,ssr,df_diff,ss_diff,F,Pr(>F)
0,998.0,3.812896e+06,0.0,NaN,NaN,NaN
1,995.0,2.984043e+06,3.0,828852.79888,92.124282,1.295833e-52



RQ1 Full Model Summary:
                            OLS Regression Results                            
Dep. Variable:        rating_gain_12m   R-squared:                       0.221
Model:                            OLS   Adj. R-squared:                  0.218
Method:                 Least Squares   F-statistic:                     70.63
Date:                Tue, 12 May 2026   Prob (F-statistic):           1.11e-52
Time:                        23:50:16   Log-Likelihood:                -5419.5
No. Observations:                1000   AIC:                         1.085e+04
Df Residuals:                     995   BIC:                         1.087e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const    

In [121]:
# ============================================================
# Save RQ1 outputs
# ============================================================

anova_rq1.to_csv(OUTPUT_TABLES_DIR / "rq1_nested_regression_test.csv")

rq1_summary = pd.DataFrame({
    "Metric": [
        "Null Model R-squared",
        "Full Model R-squared",
        "Full Model Adjusted R-squared",
        "F-statistic",
        "p-value"
    ],
    "Value": [
        model_null.rsquared,
        model_full.rsquared,
        model_full.rsquared_adj,
        anova_rq1.loc[1, "F"],
        anova_rq1.loc[1, "Pr(>F)"]
    ]
})

rq1_summary.to_csv(OUTPUT_TABLES_DIR / "rq1_summary.csv", index=False)
display(rq1_summary)

,Metric,Value
0,Null Model R-squared,4.807759e-03
1,Full Model R-squared,2.211441e-01
2,Full Model Adjusted R-squared,2.180130e-01
3,F-statistic,9.212428e+01
4,p-value,1.295833e-52


For RQ1, the null model using only current rating had: R² = 0.0048
And, The full baseline model using current rating plus activity variables had:
R² = 0.2211
Adjusted R² = 0.2180
F = 92.12
p = 1.30e-52

So, reject H0 for RQ1. Current rating alone explains almost nothing about future 12-month rating gain, but adding games played, active months, and inactive months significantly improves explanatory power.

For RQ1, a nested OLS regression compared a rating-only model with a multivariate baseline model including rating at cutoff, total games played, active months, and inactive months. The rating-only model explained very little variance in 12-month rating gain, R² = .005. The full model improved explanatory power substantially, R² = .221, adjusted R² = .218. The nested model comparison was statistically significant, F(3, 995) = 92.12, p < .001. Therefore, the null hypothesis was rejected. Rating-history and activity variables significantly improve prediction of 12-month Standard rating gain beyond current rating alone.



RQ2 hypothesis testing

RQ2 tests whether opponent-strength and tournament-performance variables improve prediction beyond baseline FIDE features.

In [124]:
# ============================================================
# Step 4: RQ2 Hypothesis Test
# Baseline FIDE-only model vs enriched opponent-strength model
# ============================================================

baseline_features = [
    "rating_start",
    "total_games_12m",
    "active_months_12m",
    "inactive_months_12m"
]

enriched_features = [
    "calc_games_12m",
    "calc_tournaments_12m",
    "avg_opponent_rating_12m",
    "score_percentage_12m",
    "games_vs_2000plus_12m",
    "total_k_chg_12m"
]

# Keep only enriched features that exist
enriched_features = [col for col in enriched_features if col in model_df.columns]

rq2_data = model_df[model_df["has_calculation_data"] == 1].copy()

rq2_cols = [target_col] + baseline_features + enriched_features
rq2_data = rq2_data[rq2_cols].dropna().copy()

print("RQ2 sample size:", rq2_data.shape)
print("RQ2 enriched features:", enriched_features)

y = rq2_data[target_col]

# Baseline model
X_base = sm.add_constant(rq2_data[baseline_features])
model_base = sm.OLS(y, X_base).fit()

# Enriched model
all_features = baseline_features + enriched_features
X_enriched = sm.add_constant(rq2_data[all_features])
model_enriched = sm.OLS(y, X_enriched).fit()

anova_rq2 = anova_lm(model_base, model_enriched)

print("RQ2 Baseline R-squared:", round(model_base.rsquared, 4))
print("RQ2 Enriched R-squared:", round(model_enriched.rsquared, 4))
print("RQ2 Enriched Adjusted R-squared:", round(model_enriched.rsquared_adj, 4))

print("\nRQ2 Nested Regression F-test:")
display(anova_rq2)

print("\nRQ2 Enriched Model Summary:")
print(model_enriched.summary())

RQ2 sample size: (629, 11)
RQ2 enriched features: ['calc_games_12m', 'calc_tournaments_12m', 'avg_opponent_rating_12m', 'score_percentage_12m', 'games_vs_2000plus_12m', 'total_k_chg_12m']
RQ2 Baseline R-squared: 0.2248
RQ2 Enriched R-squared: 0.3563
RQ2 Enriched Adjusted R-squared: 0.3459

RQ2 Nested Regression F-test:


,df_resid,ssr,df_diff,ss_diff,F,Pr(>F)
0,624.0,2.486716e+06,0.0,NaN,NaN,NaN
1,618.0,2.064796e+06,6.0,421920.823272,21.047044,1.601435e-22



RQ2 Enriched Model Summary:
                            OLS Regression Results                            
Dep. Variable:        rating_gain_12m   R-squared:                       0.356
Model:                            OLS   Adj. R-squared:                  0.346
Method:                 Least Squares   F-statistic:                     34.21
Date:                Tue, 12 May 2026   Prob (F-statistic):           4.89e-53
Time:                        23:51:41   Log-Likelihood:                -3438.8
No. Observations:                 629   AIC:                             6900.
Df Residuals:                     618   BIC:                             6949.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                              coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------

In [126]:
# ============================================================
# Save RQ2 outputs
# ============================================================

anova_rq2.to_csv(OUTPUT_TABLES_DIR / "rq2_baseline_vs_enriched_test.csv")

rq2_summary = pd.DataFrame({
    "Metric": [
        "Baseline Model R-squared",
        "Enriched Model R-squared",
        "Enriched Model Adjusted R-squared",
        "F-statistic",
        "p-value"
    ],
    "Value": [
        model_base.rsquared,
        model_enriched.rsquared,
        model_enriched.rsquared_adj,
        anova_rq2.loc[1, "F"],
        anova_rq2.loc[1, "Pr(>F)"]
    ]
})

rq2_summary.to_csv(OUTPUT_TABLES_DIR / "rq2_summary.csv", index=False)
display(rq2_summary)

,Metric,Value
0,Baseline Model R-squared,2.247938e-01
1,Enriched Model R-squared,3.563229e-01
2,Enriched Model Adjusted R-squared,3.459074e-01
3,F-statistic,2.104704e+01
4,p-value,1.601435e-22


For RQ2, the baseline model on the enriched subset had: R² = 0.2248
The enriched model had:
R² = 0.3563
Adjusted R² = 0.3459
F = 21.05
p = 1.60e-22

So, reject H0 for RQ2. Opponent-strength and tournament-performance variables add significant predictive value beyond baseline FIDE rating-history variables.

For RQ2, a nested OLS regression was conducted on the 629-player enriched subset with FIDE calculation-page data. The baseline model included rating at cutoff, games played, active months, and inactive months. The enriched model added calculation-page features, including extracted game count, tournament count, average opponent rating, score percentage, games against 2000+ opponents, and total K-adjusted rating-change contribution. The enriched model improved R² from .225 to .356, with adjusted R² = .346. The nested comparison was statistically significant, F(6, 618) = 21.05, p < .001. Therefore, the null hypothesis was rejected. Opponent-strength and tournament-performance variables significantly improve prediction beyond baseline FIDE monthly rating-history variables.


Interpret RQ1 and RQ2

For RQ1:
If p < .05:
Reject H0. The multivariate model significantly improves prediction over current rating alone.

If p >= .05:
Do not reject H0. The additional predictors did not significantly improve the linear model.

In [ ]:
For RQ2:
If p < .05:
Reject H0. Opponent-strength and tournament-performance variables significantly improve prediction.

If p >= .05:
Do not reject H0. Enriched features did not significantly improve the model in this sample.

regression model comparison

In [129]:
# ============================================================
# Regression model comparison for 12-month rating gain
# ============================================================

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np
import pandas as pd

target_col = "rating_gain_12m"

baseline_features = [
    "rating_start",
    "total_games_12m",
    "active_months_12m",
    "inactive_months_12m"
]

enriched_features = [
    "calc_games_12m",
    "calc_tournaments_12m",
    "avg_opponent_rating_12m",
    "score_percentage_12m",
    "games_vs_2000plus_12m",
    "total_k_chg_12m"
]

categorical_candidates = ["Sex", "Title", "rating_band"]
categorical_features = [col for col in categorical_candidates if col in model_df.columns]

regression_features = baseline_features + enriched_features + categorical_features

# Use enriched subset for enriched model comparison
reg_data = model_df[model_df["has_calculation_data"] == 1].copy()
reg_data = reg_data[[target_col] + regression_features].dropna(subset=[target_col]).copy()

X = reg_data[regression_features]
y = reg_data[target_col]

numeric_features = X.select_dtypes(include=[np.number]).columns.tolist()
categorical_features = X.select_dtypes(exclude=[np.number]).columns.tolist()

numeric_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore"))
])

preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

models = {
    "Linear Regression": LinearRegression(),
    "Ridge Regression": Ridge(alpha=1.0),
    "Lasso Regression": Lasso(alpha=0.01, max_iter=10000),
    "Elastic Net": ElasticNet(alpha=0.01, l1_ratio=0.5, max_iter=10000),
    "Random Forest": RandomForestRegressor(
        n_estimators=300,
        random_state=42,
        min_samples_leaf=5
    ),
    "Gradient Boosting": GradientBoostingRegressor(random_state=42)
}

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

regression_results = []

for model_name, model in models.items():
    pipe = Pipeline(steps=[
        ("preprocessor", preprocessor),
        ("model", model)
    ])

    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_test)

    mae = mean_absolute_error(y_test, y_pred)

    # Compatible RMSE calculation
    mse = mean_squared_error(y_test, y_pred)
    rmse = np.sqrt(mse)

    r2 = r2_score(y_test, y_pred)

    regression_results.append({
        "Model": model_name,
        "MAE": round(mae, 2),
        "RMSE": round(rmse, 2),
        "R2": round(r2, 3)
    })

regression_results_df = (
    pd.DataFrame(regression_results)
    .sort_values("RMSE")
    .reset_index(drop=True)
)

display(regression_results_df)

# Save results
regression_results_df.to_csv(
    OUTPUT_TABLES_DIR / "regression_model_results.csv",
    index=False
)

print("Saved:", OUTPUT_TABLES_DIR / "regression_model_results.csv")

,Model,MAE,RMSE,R2
0,Random Forest,38.61,61.34,0.390
1,Gradient Boosting,41.17,64.14,0.334
2,Lasso Regression,47.36,64.65,0.323
3,Elastic Net,47.27,64.68,0.322
4,Ridge Regression,47.40,64.71,0.322
5,Linear Regression,47.70,65.20,0.311


Saved: /Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess/outputs/tables/regression_model_results.csv


This means the Random Forest model explains about 39.0% of the variation in 12-month rating gain on the test set. Its average absolute error is about 38.6 rating points, and its RMSE is about 61.3 rating points.

The tree-based models outperform linear models, which suggests that rating gain is not purely linear. Interactions such as rating level × activity, activity × opponent strength, or score percentage × opposition strength may matter.

Regression models were trained to predict 12-month Standard rating gain using the enriched feature set. Among the tested models, Random Forest achieved the best performance, with MAE = 38.61, RMSE = 61.34, and R² = .390. Gradient Boosting was the second-best model, with MAE = 41.17, RMSE = 64.14, and R² = .334. Linear and regularized regression models performed lower, with R² values between .311 and .323. These results suggest that non-linear relationships and feature interactions are important in predicting youth chess rating gain. The Random Forest model is therefore the strongest candidate regression model at this stage, although linear models remain useful for interpretability and hypothesis testing.

In [133]:
# ============================================================
# Feature importance from best Random Forest model
# ============================================================

best_model = RandomForestRegressor(
    n_estimators=300,
    random_state=42,
    min_samples_leaf=5
)

best_pipe = Pipeline(steps=[
    ("preprocessor", preprocessor),
    ("model", best_model)
])

best_pipe.fit(X_train, y_train)

# Get transformed feature names
feature_names = best_pipe.named_steps["preprocessor"].get_feature_names_out()

importances = best_pipe.named_steps["model"].feature_importances_

feature_importance_df = (
    pd.DataFrame({
        "Feature": feature_names,
        "Importance": importances
    })
    .sort_values("Importance", ascending=False)
    .reset_index(drop=True)
)

display(feature_importance_df.head(20))

# Save feature importance
feature_importance_df.to_csv(
    OUTPUT_TABLES_DIR / "random_forest_feature_importance.csv",
    index=False
)

print("Saved:", OUTPUT_TABLES_DIR / "random_forest_feature_importance.csv")

,Feature,Importance
0,num__total_games_12m,0.437768
1,num__rating_start,0.137784
2,num__avg_opponent_rating_12m,0.134178
3,num__score_percentage_12m,0.069145
4,num__active_months_12m,0.065256
5,num__total_k_chg_12m,0.059483
6,num__calc_games_12m,0.040964
7,num__inactive_months_12m,0.021702
8,num__calc_tournaments_12m,0.021377
9,cat__Sex_M,0.002814


Saved: /Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess/outputs/tables/random_forest_feature_importance.csv


Observations:
Feature-importance analysis from the Random Forest model indicates that total rated games in the feature window was the most influential predictor of 12-month rating gain, with an importance score of .438. Starting rating was the second most important feature (.138), followed closely by average opponent rating (.134). Score percentage (.069), active months (.065), and total K-adjusted rating change (.059) also contributed to prediction. These results support the central finding that future rating gain is better explained by a combination of activity intensity, current rating context, strength of opposition, and recent performance quality than by current rating alone. Gender, title status, and rating-band dummy variables contributed very little to the Random Forest model, suggesting that they should be treated mainly as descriptive or control variables rather than primary talent-identification signals.


classification for crossing 1800 and 2000.

In [137]:
# ============================================================
# Create milestone-crossing targets and check class balance
# ============================================================

# Check available rating columns
print("Available rating-related columns:")
print([col for col in model_df.columns if "rating" in col.lower()])

# Create crossing targets if rating_end is available
if "rating_end" in model_df.columns:
    model_df["cross_1800_12m"] = (model_df["rating_end"] >= 1800).astype(int)
    model_df["cross_2000_12m"] = (model_df["rating_end"] >= 2000).astype(int)
else:
    raise ValueError("rating_end column not found. Please check your April 2026 rating column name.")

print("cross_1800_12m:")
display(model_df["cross_1800_12m"].value_counts())

print("\ncross_2000_12m:")
display(model_df["cross_2000_12m"].value_counts())

Available rating-related columns:
['rating_start', 'rating_end', 'rating_gain_12m', 'rating_band', 'avg_opponent_rating_12m', 'max_opponent_rating_12m']
cross_1800_12m:


cross_1800_12m
0    960
1     40
Name: count, dtype: int64


cross_2000_12m:


cross_2000_12m
0    985
1     15
Name: count, dtype: int64


| Target           | Negative class | Positive class | Positive rate |
| ---------------- | -------------: | -------------: | ------------: |
| `cross_1800_12m` |            960 |             40 |          4.0% |
| `cross_2000_12m` |            985 |             15 |          1.5% |

For cross_2000_12m, only 15 positives is very small, so treat it as exploratory. cross_1800_12m is more useful.




In [ ]:
classification for cross_1800_12m

In [141]:
# ============================================================
# Classification model for crossing 1800 within 12 months
# ============================================================

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    roc_auc_score, average_precision_score, confusion_matrix,
    classification_report
)
import pandas as pd
import numpy as np

class_target = "cross_1800_12m"

classification_features = [
    "rating_start",
    "total_games_12m",
    "active_months_12m",
    "inactive_months_12m",
    "calc_games_12m",
    "calc_tournaments_12m",
    "avg_opponent_rating_12m",
    "score_percentage_12m",
    "games_vs_2000plus_12m",
    "total_k_chg_12m",
    "Sex",
    "Title",
    "rating_band"
]

classification_features = [col for col in classification_features if col in model_df.columns]

class_df = model_df[[class_target] + classification_features].dropna(subset=[class_target]).copy()

X = class_df[classification_features]
y = class_df[class_target]

numeric_features = X.select_dtypes(include=[np.number]).columns.tolist()
categorical_features = X.select_dtypes(exclude=[np.number]).columns.tolist()

numeric_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore"))
])

preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

class_models = {
    "Logistic Regression": LogisticRegression(max_iter=1000, class_weight="balanced"),
    "Random Forest Classifier": RandomForestClassifier(
        n_estimators=300,
        random_state=42,
        class_weight="balanced",
        min_samples_leaf=3
    ),
    "Gradient Boosting Classifier": GradientBoostingClassifier(random_state=42)
}

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

classification_results = []
confusion_outputs = {}

for model_name, model in class_models.items():
    pipe = Pipeline(steps=[
        ("preprocessor", preprocessor),
        ("model", model)
    ])
    
    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_test)
    
    if hasattr(pipe.named_steps["model"], "predict_proba"):
        y_prob = pipe.predict_proba(X_test)[:, 1]
    else:
        y_prob = None
    
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, zero_division=0)
    recall = recall_score(y_test, y_pred, zero_division=0)
    f1 = f1_score(y_test, y_pred, zero_division=0)
    
    roc_auc = roc_auc_score(y_test, y_prob) if y_prob is not None else np.nan
    pr_auc = average_precision_score(y_test, y_prob) if y_prob is not None else np.nan
    
    cm = confusion_matrix(y_test, y_pred)
    confusion_outputs[model_name] = cm
    
    classification_results.append({
        "Model": model_name,
        "Accuracy": round(accuracy, 3),
        "Precision": round(precision, 3),
        "Recall": round(recall, 3),
        "F1": round(f1, 3),
        "ROC_AUC": round(roc_auc, 3),
        "PR_AUC": round(pr_auc, 3)
    })

classification_1800_results_df = (
    pd.DataFrame(classification_results)
    .sort_values("PR_AUC", ascending=False)
    .reset_index(drop=True)
)

display(classification_1800_results_df)

print("\nConfusion matrices:")
for model_name, cm in confusion_outputs.items():
    print("\n", model_name)
    print(cm)

# Save results
classification_1800_results_df.to_csv(
    OUTPUT_TABLES_DIR / "classification_cross_1800_results.csv",
    index=False
)

print("Saved:", OUTPUT_TABLES_DIR / "classification_cross_1800_results.csv")

,Model,Accuracy,Precision,Recall,F1,ROC_AUC,PR_AUC
0,Logistic Regression,0.980,0.700,0.875,0.778,0.986,0.908
1,Random Forest Classifier,0.985,0.857,0.750,0.800,0.898,0.865
2,Gradient Boosting Classifier,0.980,0.750,0.750,0.750,0.924,0.817



Confusion matrices:

 Logistic Regression
[[189   3]
 [  1   7]]

 Random Forest Classifier
[[191   1]
 [  2   6]]

 Gradient Boosting Classifier
[[190   2]
 [  2   6]]
Saved: /Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess/outputs/tables/classification_cross_1800_results.csv


Observations:
| Model               | Accuracy | Precision | Recall |   F1 | ROC-AUC | PR-AUC |
| ------------------- | -------: | --------: | -----: | ---: | ------: | -----: |
| Logistic Regression |     .980 |      .700 |   .875 | .778 |    .986 |   .908 |
| Random Forest       |     .985 |      .857 |   .750 | .800 |    .898 |   .865 |
| Gradient Boosting   |     .980 |      .750 |   .750 | .750 |    .924 |   .817 |



#### Observations:
For the cross_1800_12m classification task, the target was highly imbalanced, with only 40 of 1,000 players reaching or exceeding 1800 during the outcome window. Therefore, precision, recall, F1-score, ROC-AUC, and PR-AUC were emphasized over accuracy. Logistic Regression achieved the highest recall (.875) and PR-AUC (.908), correctly identifying 7 of 8 positive cases in the test set. Random Forest achieved the highest F1-score (.800) and precision (.857), indicating fewer false positives but slightly lower recall. These results suggest that milestone-crossing prediction is feasible, but model selection should depend on the practical objective: maximizing recall for scouting or maximizing precision for a narrower shortlist.

In [148]:
class_target = "cross_2000_12m"

In [150]:
# ============================================================
# Classification model for crossing 2000 within 12 months
# Exploratory because positive class is very small
# ============================================================

class_target = "cross_2000_12m"

classification_features = [
    "rating_start",
    "total_games_12m",
    "active_months_12m",
    "inactive_months_12m",
    "calc_games_12m",
    "calc_tournaments_12m",
    "avg_opponent_rating_12m",
    "score_percentage_12m",
    "games_vs_2000plus_12m",
    "total_k_chg_12m",
    "Sex",
    "Title",
    "rating_band"
]

classification_features = [col for col in classification_features if col in model_df.columns]

class_df = model_df[[class_target] + classification_features].dropna(subset=[class_target]).copy()

X = class_df[classification_features]
y = class_df[class_target]

print("Target distribution:")
print(y.value_counts())

numeric_features = X.select_dtypes(include=[np.number]).columns.tolist()
categorical_features = X.select_dtypes(exclude=[np.number]).columns.tolist()

numeric_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore"))
])

preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

class_models = {
    "Logistic Regression": LogisticRegression(max_iter=1000, class_weight="balanced"),
    "Random Forest Classifier": RandomForestClassifier(
        n_estimators=300,
        random_state=42,
        class_weight="balanced",
        min_samples_leaf=3
    ),
    "Gradient Boosting Classifier": GradientBoostingClassifier(random_state=42)
}

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

classification_results = []
confusion_outputs_2000 = {}

for model_name, model in class_models.items():
    pipe = Pipeline(steps=[
        ("preprocessor", preprocessor),
        ("model", model)
    ])

    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_test)

    if hasattr(pipe.named_steps["model"], "predict_proba"):
        y_prob = pipe.predict_proba(X_test)[:, 1]
    else:
        y_prob = None

    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, zero_division=0)
    recall = recall_score(y_test, y_pred, zero_division=0)
    f1 = f1_score(y_test, y_pred, zero_division=0)

    roc_auc = roc_auc_score(y_test, y_prob) if y_prob is not None else np.nan
    pr_auc = average_precision_score(y_test, y_prob) if y_prob is not None else np.nan

    cm = confusion_matrix(y_test, y_pred)
    confusion_outputs_2000[model_name] = cm

    classification_results.append({
        "Model": model_name,
        "Accuracy": round(accuracy, 3),
        "Precision": round(precision, 3),
        "Recall": round(recall, 3),
        "F1": round(f1, 3),
        "ROC_AUC": round(roc_auc, 3),
        "PR_AUC": round(pr_auc, 3)
    })

classification_2000_results_df = (
    pd.DataFrame(classification_results)
    .sort_values("PR_AUC", ascending=False)
    .reset_index(drop=True)
)

display(classification_2000_results_df)

print("\nConfusion matrices:")
for model_name, cm in confusion_outputs_2000.items():
    print("\n", model_name)
    print(cm)

# Save results
classification_2000_results_df.to_csv(
    OUTPUT_TABLES_DIR / "classification_cross_2000_results.csv",
    index=False
)

print("Saved:", OUTPUT_TABLES_DIR / "classification_cross_2000_results.csv")

Target distribution:
cross_2000_12m
0    985
1     15
Name: count, dtype: int64


,Model,Accuracy,Precision,Recall,F1,ROC_AUC,PR_AUC
0,Logistic Regression,0.975,0.333,0.667,0.444,0.983,0.744
1,Gradient Boosting Classifier,0.995,1.000,0.667,0.800,0.831,0.672
2,Random Forest Classifier,0.990,0.667,0.667,0.667,0.976,0.456



Confusion matrices:

 Logistic Regression
[[193   4]
 [  1   2]]

 Random Forest Classifier
[[196   1]
 [  1   2]]

 Gradient Boosting Classifier
[[197   0]
 [  1   2]]
Saved: /Users/arunkumar/Desktop/Arun/Arun/DBA-Walsh/SecondYr/Capstone/ML_Driven_TalentIdentification_In_Youth_Chess/outputs/tables/classification_cross_2000_results.csv


Caution: since only 15 players crossed 2000, we treat cross_2000_12m results as exploratory, not as strong final evidence.

#### Observations:
Only 15 of 1,000 players crossed 2000, so this is a rare-event classification problem. The test set likely had only 3 positive cases, so results should be treated as exploratory.
| Model               | Accuracy | Precision | Recall |   F1 | ROC-AUC | PR-AUC |
| ------------------- | -------: | --------: | -----: | ---: | ------: | -----: |
| Logistic Regression |     .975 |      .333 |   .667 | .444 |    .983 |   .744 |
| Gradient Boosting   |     .995 |     1.000 |   .667 | .800 |    .831 |   .672 |
| Random Forest       |     .990 |      .667 |   .667 | .667 |    .976 |   .456 |

Gradient Boosting gives the best F1-score: F1 = .800 ; Precision = 1.000 ; Recall = .667
confusion matrix : [[197   0] [  1   2]]

So it correctly identified 2 of 3 actual 2000-crossers in the test set and made zero false-positive predictions.

However, because the positive class is extremely small, we should not overclaim. This result is promising, but not as statistically stable as the 1800-crossing model.

The cross_2000_12m target was more severely imbalanced than the 1800 milestone, with only 15 of 1,000 players reaching or exceeding 2000 during the outcome window. Therefore, the 2000-threshold classification results are treated as exploratory. Gradient Boosting achieved the strongest F1-score (.800), with precision = 1.000 and recall = .667, correctly identifying 2 of 3 positive cases in the test set without false positives. Logistic Regression achieved the highest PR-AUC (.744) and ROC-AUC (.983), suggesting strong ranking ability, but its precision was lower. Because of the very small number of positive cases, these results should be validated on a larger sample before drawing firm conclusions about 2000-level milestone prediction. This should be considered next step.

#### Overall:
Overall, milestone-crossing models performed well despite class imbalance. The 1800-crossing task was more reliable because it had 40 positive cases, while the 2000-crossing task had only 15 positive cases and should be considered exploratory. For scouting-oriented use, recall and PR-AUC are more important than accuracy because missing a high-potential player may be more costly than reviewing a small number of false positives.

developmental trajectory labels

| Trajectory label    | Rule used                                              | Expanded meaning                                                                                                                                                                                                                                                                                                                           |
| ------------------- | ------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **High-growth**     | `rating_gain_12m >= 75`                                | Player showed a large rating jump over the next 12 months. This group represents accelerated improvers whose rating gain is clearly above normal year-to-year movement. These players may be strong candidates for closer coaching attention, tournament planning, or talent-development support.                                          |
| **Stable progress** | `20 <= rating_gain_12m < 75`                           | Player improved meaningfully but not at an exceptional rate. This group reflects steady development, where the player is moving upward but not showing the sharp jump seen in the high-growth group. These players may benefit from continued structured training and consistent tournament exposure.                                      |
| **Plateau**         | `-20 <= rating_gain_12m < 20`                          | Player’s rating remained broadly stable over the 12-month outcome window. Small gains or losses in this range are treated as normal rating fluctuation rather than clear improvement or decline. This group may include players who are maintaining strength but not yet showing measurable rating progress.                               |
| **Decline**         | `rating_gain_12m < -20`                                | Player lost more than 20 rating points over the outcome window. This suggests negative rating movement beyond minor fluctuation. The label does not mean permanent decline, but it may indicate performance difficulty, rating volatility, stronger opposition, reduced preparation, or other factors that require closer review.          |
| **Inactivity risk** | `inactive_months_12m >= 10` and `rating_gain_12m <= 0` | Player was inactive for most of the feature window and did not show positive rating gain. This group is intended to capture disengagement or low-competition exposure risk. It does not imply low ability; rather, it identifies players whose lack of rated activity limits observable development and may reduce future rating progress. |


These trajectory labels are operational definitions created for this study. They are not official FIDE categories and should not be interpreted as permanent judgments about a player’s ability. The labels are intended to support modelling, subgroup comparison, and practical interpretation of different development patterns.

rating_gain_12m:
mean     10.39
std      61.93
min    -217.00
25%     -10.00
50%       0.00
75%      12.25
max     404.00

The trajectory cutoffs were defined as operational thresholds based on the observed distribution of 12-month rating gain and practical interpretability. Because the median gain was 0 and the 75th percentile was approximately 12 rating points, small rating changes were treated as normal fluctuation. A ±20 point band was therefore used to define plateau, while gains of 20–74 points were treated as stable progress. Gains of 75 or more points were classified as high-growth because they represent a substantially larger improvement than the typical player in the sample. Players with 10 or more inactive months and no positive gain were labelled inactivity risk to capture low activity and limited observable development. These thresholds are project-defined and should be validated through sensitivity analysis in the final model.

In [166]:
# ============================================================
# Create developmental trajectory labels for RQ4
# ============================================================

def assign_trajectory(row):
    gain = row["rating_gain_12m"]
    inactive = row.get("inactive_months_12m", np.nan)
    
    if pd.isna(gain):
        return np.nan
    
    if inactive >= 10 and gain <= 0:
        return "Inactivity risk"
    elif gain >= 75:
        return "High-growth"
    elif gain >= 20:
        return "Stable progress"
    elif -20 <= gain < 20:
        return "Plateau"
    else:
        return "Decline"

model_df["trajectory_label"] = model_df.apply(assign_trajectory, axis=1)

print("Trajectory label distribution:")
print(model_df["trajectory_label"].value_counts(dropna=False))

Trajectory label distribution:
trajectory_label
Plateau            339
Inactivity risk    294
Decline            136
High-growth        123
Stable progress    108
Name: count, dtype: int64


##### Notes:
This trajectory distribution is good enough for multiclass modelling. The classes are not perfectly balanced, but none is extremely tiny.

The trajectory-label distribution was sufficiently usable for multiclass modelling. Plateau was the largest group, with 339 players, followed by inactivity risk with 294 players. Decline, high-growth, and stable progress were smaller but still represented by more than 100 players each. Because the classes remain moderately imbalanced, macro-F1 and class-level recall were emphasized instead of overall accuracy alone.


In [187]:

value = 142

players = model_df.loc[
    model_df["rating_gain_12m"] > value,
    [
        "ID Number",
        "Name",
        "Fed",
        "Sex",
        "Title",
        "Birth_Year",
        "rating_start",
        "rating_end",
        "rating_gain_12m",
        "total_games_12m",
        "active_months_12m",
        "inactive_months_12m",
        "rating_band",
        "has_calculation_data",
        "avg_opponent_rating_12m",
        "score_percentage_12m",
    ]
]

print("Number of players with rating gain/drop =", value, ":", len(players))
display(players)

Number of players with rating gain/drop = 142 : 49


,ID Number,Name,Fed,Sex,Title,Birth_Year,rating_start,rating_end,rating_gain_12m,total_games_12m,active_months_12m,inactive_months_12m,rating_band,has_calculation_data,avg_opponent_rating_12m,score_percentage_12m
15,25158910,Abhineeth Bhat,IND,M,NaN,2011.0,1651,1799,148,104,10,2,1600-1799,1,1632.344444,0.455556
27,33341800,Shaurya Chaudhry,IND,M,NaN,2010.0,1808,2002,194,26,2,10,1800-1999,1,1703.527778,0.486111
33,48726621,Subhradeep Roy,IND,M,NaN,2011.0,1566,1750,184,38,4,8,1400-1599,1,1623.069767,0.476744
34,429040531,Barnil Baishya,IND,M,NaN,2015.0,1408,1601,193,14,2,7,1400-1599,1,1538.937500,0.312500
48,33477434,Logidhashrath R,IND,M,NaN,2010.0,1508,1656,148,15,3,9,1400-1599,1,1615.382353,0.382353
60,429024722,Aarya Agarwal,IND,F,NaN,2014.0,1433,1576,143,15,3,9,1400-1599,1,1527.357143,0.178571
67,33490716,Veer Kumar,IND,M,NaN,2013.0,1602,1776,174,51,6,6,1600-1799,1,1635.473684,0.473684
71,429096979,Aarush R,IND,M,NaN,2015.0,1434,1599,165,27,4,7,1400-1599,1,1654.333333,0.033333
82,547009772,Siva Kumar,IND,M,NaN,2008.0,1401,1573,172,38,3,0,1400-1599,0,NaN,NaN
109,343444890,"Raposo, Maxwell",IND,M,NaN,2012.0,1717,1974,257,129,10,2,1600-1799,1,1598.783784,0.560811


multiclass trajectory classification

In [ ]:
# ============================================================
# Multiclass classification for developmental trajectory labels
# ============================================================

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.metrics import (
    accuracy_score,
    f1_score,
    precision_score,
    recall_score,
    classification_report,
    confusion_matrix
)
import pandas as pd
import numpy as np

trajectory_target = "trajectory_label"

trajectory_features = [
    "rating_start",
    "total_games_12m",
    "active_months_12m",
    "inactive_months_12m",
    "calc_games_12m",
    "calc_tournaments_12m",
    "avg_opponent_rating_12m",
    "score_percentage_12m",
    "games_vs_2000plus_12m",
    "total_k_chg_12m",
    "Sex",
    "Title",
    "rating_band"
]

trajectory_features = [col for col in trajectory_features if col in model_df.columns]

traj_df = model_df[[trajectory_target] + trajectory_features].dropna(subset=[trajectory_target]).copy()

X = traj_df[trajectory_features]
y = traj_df[trajectory_target]

numeric_features = X.select_dtypes(include=[np.number]).columns.tolist()
categorical_features = X.select_dtypes(exclude=[np.number]).columns.tolist()

numeric_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("onehot", OneHotEncoder(handle_unknown="ignore"))
])

preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features)
    ]
)

trajectory_models = {
    "Logistic Regression": LogisticRegression(
        max_iter=2000,
        class_weight="balanced"
    ),
    "Random Forest Classifier": RandomForestClassifier(
        n_estimators=300,
        random_state=42,
        class_weight="balanced",
        min_samples_leaf=3
    ),
    "Gradient Boosting Classifier": GradientBoostingClassifier(random_state=42)
}
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

trajectory_results = []
trajectory_reports = {}
trajectory_confusion_matrices = {}

for model_name, model in trajectory_models.items():
    pipe = Pipeline(steps=[
        ("preprocessor", preprocessor),
        ("model", model)
    ])
    
    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_test)
    
    accuracy = accuracy_score(y_test, y_pred)
    macro_precision = precision_score(y_test, y_pred, average="macro", zero_division=0)
    macro_recall = recall_score(y_test, y_pred, average="macro", zero_division=0)
    macro_f1 = f1_score(y_test, y_pred, average="macro", zero_division=0)
    weighted_f1 = f1_score(y_test, y_pred, average="weighted", zero_division=0)
    
    trajectory_results.append({
        "Model": model_name,
        "Accuracy": round(accuracy, 3),
        "Macro_Precision": round(macro_precision, 3),
        "Macro_Recall": round(macro_recall, 3),
        "Macro_F1": round(macro_f1, 3),
        "Weighted_F1": round(weighted_f1, 3)
    })
    
    trajectory_reports[model_name] = classification_report(
        y_test,
        y_pred,
        zero_division=0,
        output_dict=True
    )
    
    trajectory_confusion_matrices[model_name] = confusion_matrix(
        y_test,
        y_pred,
        labels=sorted(y.unique())
    )

trajectory_results_df = (
    pd.DataFrame(trajectory_results)
    .sort_values("Macro_F1", ascending=False)
    .reset_index(drop=True)
)

display(trajectory_results_df)

print("\nClass labels used in confusion matrix:")
print(sorted(y.unique()))

print("\nConfusion matrices:")
for model_name, cm in trajectory_confusion_matrices.items():
    print("\n", model_name)
    print(cm)

# Save summary results
trajectory_results_df.to_csv(
    OUTPUT_TABLES_DIR / "trajectory_classification_results.csv",
    index=False
)

print("Saved:", OUTPUT_TABLES_DIR / "trajectory_classification_results.csv")

| Model                        | Accuracy | Macro Precision | Macro Recall | Macro F1 | Weighted F1 |
| ---------------------------- | -------: | --------------: | -----------: | -------: | ----------: |
| Gradient Boosting Classifier |     .760 |            .703 |         .667 |     .674 |        .744 |
| Random Forest Classifier     |     .740 |            .659 |         .667 |     .662 |        .739 |
| Logistic Regression          |     .655 |            .571 |         .587 |     .571 |        .661 |

The Gradient Boosting Classifier performed best overall: Accuracy = .760; Macro-F1 = .674 ; Weighted-F1 = .744

This means the model correctly classified about 76% of players into the five developmental trajectory groups. Macro-F1 is more important than accuracy here because the classes are moderately imbalanced. A macro-F1 of .674 indicates that the model is learning meaningful patterns across all trajectory groups, not only the largest classes.

For RQ4, developmental trajectory labels were created using operational thresholds based on 12-month rating gain and inactivity. The five classes were decline, high-growth, inactivity risk, plateau, and stable progress. The class distribution was moderately imbalanced but usable for multiclass modelling. Gradient Boosting achieved the best trajectory-classification performance, with accuracy = .760, macro-F1 = .674, and weighted-F1 = .744. Random Forest was close behind, with accuracy = .740 and macro-F1 = .662, while Logistic Regression performed lower, with macro-F1 = .571. These results suggest that developmental trajectories can be classified with meaningful accuracy, especially when using non-linear models. However, class-level inspection shows that stable progress is harder to distinguish from plateau and high-growth, so trajectory labels should be interpreted as decision-support categories rather than fixed judgments about players.

RQ4 Hypothesis
H0: Models do not classify trajectories better than a naïve baseline.
H1: Models classify trajectories better than a naïve baseline.

The null hypothesis for RQ4 was rejected in practical modelling terms because Gradient Boosting achieved accuracy = .760 and macro-F1 = .674 across five trajectory classes, substantially above the naïve majority-class baseline of 33.9% accuracy from predicting all players as Plateau. This indicates that the selected features provide meaningful information for developmental trajectory classification.

Caution: Because trajectory labels are project-defined operational categories, the classification results should be interpreted as a proof of modelling feasibility rather than as official or permanent player classifications.